src/kotlin/kage/crypto/stream/ArmorInputStream.kt

```diff
@@ -0,0 +1,104 @@
+/**
+ * Copyright 2023 The kage Authors. All rights reserved. Use of this source code is governed by
+ * either an Apache 2.0 or MIT license at your discretion, that can be found in the LICENSE-APACHE
+ * or LICENSE-MIT files respectively.
+ */
+package kage.crypto.stream
+
+import java.io.InputStream
+import java.util.*
+import kage.errors.ArmorCodingException
+
+internal class ArmorInputStream(src: InputStream) : InputStream() {
+  // Holds already decoded bytes
+  private val unread = ByteArray(BytesPerLine)
+  private var unreadSize = 0
+  private var unreadOffset = 0
+
+  private val srcScanner = src.bufferedReader()
+
+  private var started = false
+  private var isEOF = false
+
+  private val maxWhitespace = 1024
+
+  override fun read(): Int {
+    if (unreadOffset < unreadSize) return (unread[unreadOffset++].toInt() and 0xff)
+
+    if (isEOF) return -1
+
+    if (!started) drainLeading()
+
+    val line = srcScanner.readLine()
+
+    if (line == Footer) {
+      drainTrailing()
+      isEOF = true
+    }
+
+    if (line.length > ColumnsPerLine) {
+      throw ArmorCodingException("column limit exceeded")
+    }
+
+    val decodedLine = Base64.getDecoder().decode(line)
```

Comment:
Use the `String.decodeBase64()` method from `kage.utils.Extensions`. It makes sure that the Base64 string is canonical before decoding.

Reply:
I changed this, but a test now fails. I am a bit confused by the spec https://github.com/C2SP/C2SP/blob/main/age.md#ascii-armor

It states:

```
Note that this encoding employs base64 with = padding characters, unlike the rest of this document.
```

So, `=` is allowed, but canonical no? or yes, unlike the rest of that document?

Reply:
The same doc also mentions this

```
Note that ASCII armored files are malleable unless care is taken to reject any data before and after the PEM encoding, a strict PEM parser is used, and canonical base64 is enforced. age implementations SHOULD reject non-canonical ASCII armor encodings except for whitespace before and after the PEM block
```

So, I'm assuming the only difference is that padding characters are allowed now and we still need to make sure that the PEM block is canonical.

Reply:
@simao I think I figured out what the issue is, we need to check the complete PEM encoding to verify if it is canonical or not. Once that is done, we can decode individual lines (in ArmorInputStream) without checking if they are canonical or not.

Reply:
But that is not what the age.go implementation does?

I does validate PEM, but it does so line by line, and validates trailing and leading whitespace. Which we are also doing here.

What we are not doing is to validate that each base64 line is padded with 0 bytes. Which is what the base64 decoder provided by the golang std lib does when using strict mode. This is the decoder that age.go uses, but there is no such mode in java base64.

Reply:
I think I got it now, see latest changes.

The encoder now doesn't use padding, which produces canonical base64.

The decoder uses `kage.utils.decodeBase64` which verifies it's canonical. I changed isCanonicalBase64 so it returns the decoded payload so we avoid decoding the input twice.

Reply:
@simao cool, I actually returned boolean from `isCanonicalBase64` just because I thought we might need it somewhere else. But yeah, returning both (isCanonical and the ByteArray) solves that 👍